modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/storage/MetagraphMessageStorage.scala

```diff
@@ -0,0 +1,40 @@
+package org.tessellation.node.shared.infrastructure.snapshot.storage
+
+import cats.Functor
+import cats.effect.Ref
+import cats.syntax.functor._
+
+import org.tessellation.schema.metagraphMessage.{AdminFeeKeyMessage, StakingMessage}
+
+trait MetagraphMessageStorage[F[_]] {
+  def get: F[(Option[AdminFeeKeyMessage], Option[StakingMessage])]
+  def setAdminFeeKeyMessage(newMessage: Option[AdminFeeKeyMessage]): F[Unit]
+  def setStakingMessage(newMessage: Option[StakingMessage]): F[Unit]
+}
+
+object MetagraphMessageStorage {
+  private type StoredTuple = (Option[AdminFeeKeyMessage], Option[StakingMessage])
+
+  def make[F[_]: Functor: Ref.Make](
+    maybeAdminFeeKeyMessage: Option[AdminFeeKeyMessage],
+    maybeStakingMessage: Option[StakingMessage]
+  ): F[MetagraphMessageStorage[F]] =
+    Ref.of[F, StoredTuple]((maybeAdminFeeKeyMessage, maybeStakingMessage)).map(make(_))
+
+  def make[F[_]: Functor: Ref.Make](valuesR: Ref[F, StoredTuple]): MetagraphMessageStorage[F] =
+    new MetagraphMessageStorage[F] {
+      def get: F[StoredTuple] =
+        valuesR.get
+
+      def setAdminFeeKeyMessage(newMessage: Option[AdminFeeKeyMessage]): F[Unit] =
+        valuesR.update {
+          case (_, maybeStaking) => (newMessage, maybeStaking)
+        }
```

Comment:
you can store both in two separate Refs 🤔 just an idea as handling tuples may be more complicated

Reply:
done